### DISCUSSION TITLE
Support for monorepos

### DISCUSSION THREAD

**Discussion by alpiquero** [mooltiverse/nyx#268] [Ideas]
Hi! First of all, congratulations for your awesome work. Nyx is really helping us with our repos versioning.

I would like to ask if you are planning to support monorepos in the future. I beliebe this will be useful for a lot of teams.

**Comment by flelli** [comment]
Hi @alpiquero , can you tell me something more about your use case?
Afaik there's no no limitations using Nyx for monorepos

**Reply by alpiquero** [reply]
Hi @flelli, we have the following structure in our monorepo:
```
package1/
package2/
package3/
```

Where the three "package" folders are the three components of our monorepo.

We want that when changes were commited to `package1` Nyx will publish a new version tag just for that component and only for that component. (Example: `package1-v1.1.0`) and want the same for `package2` and `package3`.

That means that we will have several tags published for every component of our monorepo that should be independent from each other, so we will be able to version each component on an independent way. If I make a commit that changes the content of `package2` and that change bumps a version for that component, we don't want to affect the version of the other two components.

This is an example of the version tags that should exist in our repo after all the components have had their version tags bumped several times:
```
package1-v1.1.0
package2-v1.2.1
package3-v3.2.0
```

We used to use [multi-semantic-release](https://github.com/dhoulb/multi-semantic-release) to achieve what I'm saying, but the `semantic-release` tool does not work well with our git workflow.

**Comment by flelli** [comment]
Hi @alpiquero ,

I understand your point now but this is out of Nyx use cases. There may be workarounds (below) but introducing this in Nyx would break some foundational concepts. For example the [release scope](https://mooltiverse.github.io/nyx/guide/user/introduction/how-nyx-works/#infer) wouldn't be consistent, and this would break the whole thing.

Moreover I'm wondering how it should behave when a commit affects files in more than one sub-repo or files in the project root, outside of any sub-repo.

My 10 cents here, IMHO, is that this use case is somehow even outside of Git's purpose, which is instead supported by [submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules). If even tagging and versioning is separate, it looks like they should be better handled as separate repos.

Nonetheless, a possible (not tested) solution could be:

- create a Nyx configuration for each sub-repo and run Nyx from there instead of the root directory. A separate configuration is required because you need separate [release prefix](https://mooltiverse.github.io/nyx/guide/user/configuration-reference/global-options/#release-prefix) and [filter tags](https://mooltiverse.github.io/nyx/guide/user/configuration-reference/release-types/#filter-tags) for each repo
- run Nyx **conditionally** in each sub-repo only when some file within the repo has been changed. In case you use GitHub Actions the [paths-filter](https://github.com/dorny/paths-filter) may come handy
